File: apps/web/src/types/system.ts
export type ServiceState = "online" | "offline";
export type WebSocketState = "connected" | "connecting" | "disconnected";
export type ProductMode = "RESEARCH_ONLY" | "FULL_SIMULATION";

export interface SystemStatus {
  api: "online";
  postgresql: ServiceState;
  redis: ServiceState;
  environment: string;
  product_mode: ProductMode;
  version: string;
  server_time: string;
  correlation_id: string;
}

export type CapabilityImplementationStatus =
  "WORKING" | "PARTIAL" | "PLACEHOLDER" | "NOT_IMPLEMENTED" | "DISABLED";
export type CapabilityReadinessStatus =
  "READY" | "MISSING" | "DISABLED" | "NOT_REQUIRED" | "UNKNOWN";
export type CapabilityConfigurationStatus =
  | CapabilityReadinessStatus
  | "FAKE"
  | "REAL_CONFIGURED"
  | "REAL_AVAILABLE"
  | "REAL_UNAVAILABLE";

export interface SystemCapability {
  module_key: string;
  implementation_status: CapabilityImplementationStatus;
  data_status: CapabilityReadinessStatus;
  configuration_status: CapabilityConfigurationStatus;
  available: boolean;
  availability?:
    | "READY"
    | "NEEDS_DATA"
    | "NEEDS_CONFIG"
    | "DEMO_ONLY"
    | "DISABLED"
    | "PARTIAL"
    | "NOT_IMPLEMENTED"
    | "AVAILABLE"
    | "DEGRADED"
    | "NOT_AVAILABLE";
  reason: string;
  required_actions: string[];
  last_success_at?: string | null;
  provider?: string | null;
  provider_status?: "READY" | "MISSING" | "DISABLED" | "UNKNOWN";
  mode?: string;
  worker_status?: "ONLINE" | "OFFLINE" | "NOT_REQUIRED";
}

export interface CapabilityDataCounts {
  instrument_count: number | null;
  market_bar_count: number | null;
  daily_market_bar_count: number | null;
  market_bar_instrument_count: number | null;
  earliest_market_bar_at: string | null;
  latest_market_bar_at: string | null;
  simulated_account_count: number | null;
  scan_run_count: number | null;
  strategy_run_count: number | null;
  strategy_experiment_count: number | null;
  information_source_count: number | null;
  information_item_count: number | null;
  market_event_count: number | null;
  ai_analysis_run_count: number | null;
  order_count: number | null;
  executable_order_count: number | null;
  fill_count: number | null;
  risk_decision_count: number | null;
  backtest_run_count: number | null;
  replay_run_count: number | null;
  trading_calendar_session_count: number | null;
  adjustment_factor_count: number | null;
  trading_status_count: number | null;
  lifecycle_event_count: number | null;
}

export interface SystemCapabilities {
  generated_at: string;
  database_reachable: boolean;
  counts: CapabilityDataCounts;
  items: SystemCapability[];
}

export interface ApiErrorEnvelope {
  error: {
    code: string;
    message: string;
    details: unknown;
    correlation_id: string;
    timestamp: string;
  };
}

File: apps/api/src/alphadesk_api/schemas/system.py
"""Health and system status response schemas."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class LiveResponse(BaseModel):
    status: Literal["alive"]
    service: str
    version: str
    timestamp: datetime


class ReadyResponse(BaseModel):
    status: Literal["ready", "not_ready"]
    postgresql: Literal["online", "offline"]
    redis: Literal["online", "offline"]
    timestamp: datetime


class SystemStatusResponse(BaseModel):
    api: Literal["online"]
    postgresql: Literal["online", "offline"]
    redis: Literal["online", "offline"]
    environment: str
    version: str
    server_time: datetime
    correlation_id: str


class CapabilityDataCountsResponse(BaseModel):
    migration_head: str | None = None
    instrument_count: int | None = None
    market_bar_count: int | None = None
    daily_market_bar_count: int | None = None
    intraday_1m_bar_count: int | None = None
    intraday_5m_bar_count: int | None = None
    intraday_15m_bar_count: int | None = None
    intraday_30m_bar_count: int | None = None
    intraday_60m_bar_count: int | None = None
    market_bar_instrument_count: int | None = None
    earliest_market_bar_at: datetime | None = None
    latest_market_bar_at: datetime | None = None
    simulated_account_count: int | None = None
    scan_run_count: int | None = None
    strategy_run_count: int | None = None
    strategy_experiment_count: int | None = None
    information_source_count: int | None = None
    information_item_count: int | None = None
    market_event_count: int | None = None
    ai_analysis_run_count: int | None = None
    order_count: int | None = None
    executable_order_count: int | None = None
    fill_count: int | None = None
    risk_decision_count: int | None = None
    backtest_run_count: int | None = None
    replay_run_count: int | None = None
    trading_calendar_session_count: int | None = None
    adjustment_factor_count: int | None = None
    trading_status_count: int | None = None
    lifecycle_event_count: int | None = None


class SystemCapabilityResponse(BaseModel):
    module_key: str
    implementation_status: Literal["WORKING", "PARTIAL", "PLACEHOLDER", "NOT_IMPLEMENTED"]
    data_status: Literal["READY", "MISSING", "DISABLED", "NOT_REQUIRED", "UNKNOWN"]
    configuration_status: Literal[
        "READY",
        "MISSING",
        "DISABLED",
        "NOT_REQUIRED",
        "UNKNOWN",
        "FAKE",
        "REAL_CONFIGURED",
        "REAL_AVAILABLE",
        "REAL_UNAVAILABLE",
    ]
    available: bool
    availability: Literal[
        "READY",
        "NEEDS_DATA",
        "NEEDS_CONFIG",
        "DEMO_ONLY",
        "DISABLED",
        "PARTIAL",
        "NOT_IMPLEMENTED",
        "AVAILABLE",
        "DEGRADED",
        "NOT_AVAILABLE",
    ]
    reason: str
    required_actions: list[str]
    last_success_at: datetime | None = None
    provider: str | None = None
    provider_status: Literal["READY", "MISSING", "DISABLED", "UNKNOWN"] = "UNKNOWN"
    mode: str
    worker_status: Literal["ONLINE", "OFFLINE", "NOT_REQUIRED"] = "NOT_REQUIRED"


class SystemCapabilitiesResponse(BaseModel):
    generated_at: datetime
    database_reachable: bool
    counts: CapabilityDataCountsResponse
    items: list[SystemCapabilityResponse]


class WebSocketMessage(BaseModel):
    type: Literal["connected", "pong", "error"]
    timestamp: datetime
    message: str | None = None
